lib/user_app_jwt/user_app_jwt.c

```diff
@@ -0,0 +1,758 @@
+{
+	int err = 0;
+	uint8_t pub_key_hash[ECDSA_SHA_256_HASH_SZ];
+	char *hdr_str = NULL;
+	cJSON *jwt_hdr = cJSON_CreateObject();
+
+	// Type: format: always "JWT"
+	if (cJSON_AddStringToObjectCS(jwt_hdr, "typ", "JWT") == NULL) {
+		return NULL;
+	}
+
+	// Algorithme: format: always "ES256"
+	if (JWT_ALG_TYPE_ES256 == alg) {
+		if (cJSON_AddStringToObjectCS(jwt_hdr, "alg", "ES256") == NULL) {
+			return NULL;
+		}
+	} else {
+		return NULL;
+	}
+
+	// Keyid: format: sha256 string
+	/* init crypto service */
+	err = crypto_init();
+	if (err) {
+		LOG_ERR("Failed to initialize PSA Crypto, error: %d", err);
+		return NULL;
+	}
+
+	/* Get kid: sha256 over public key */
+	size_t olen;
+	err = export_public_key_hash(IAK_APPLICATION_GEN1, pub_key_hash, ECDSA_SHA_256_HASH_SZ,
+				     &olen);
+	if (err) {
+		LOG_ERR("Failed to export public key, error: %d", err);
+		return NULL;
+	}
+
+	char sha256_str[ECDSA_SHA_256_SIG_SZ + 1] = {0};
+	int32_t printed_bytes = 0;
+	for (uint32_t i = 0; i < 8; i++) {
+		printed_bytes += sprintf((sha256_str + printed_bytes), "%08x",
+					 swap_endian((uint32_t)*((uint32_t *)pub_key_hash + i)));
+	}
+	sha256_str[ECDSA_SHA_256_SIG_SZ] = '\0';
+
+	if (cJSON_AddStringToObjectCS(jwt_hdr, "kid", sha256_str) == NULL) {
+		err = -ENOMEM;
+	}
+
+	if (err == 0) {
+		hdr_str = cJSON_PrintUnformatted(jwt_hdr);
+	}
+
+	cJSON_Delete(jwt_hdr);
+	jwt_hdr = NULL;
+
+	return hdr_str;
+}
+
+//---------------------------------------------------------------------------//
+static char *convert_str_to_b64_url(const char *const str)
+{
+	size_t str_len = strlen(str);
+	size_t b64_sz = BASE64_ENCODE_SZ(str_len) + 1;
+
+	char *const b64_out = calloc(b64_sz, 1);
+
+	if (b64_out) {
+		int err = base64_encode(b64_out, b64_sz, &b64_sz, str, strlen(str));
+
+		if (err) {
+			LOG_ERR("base64_encode failed, error: %d", err);
+		}
+	}
+
+	/* Convert to base64 URL */
+	base64_url_format(b64_out);
+
+	return b64_out;
+}
+
+//---------------------------------------------------------------------------//
+static char *jwt_header_payload_combine(const char *const hdr, const char *const pay)
+{
+	/* Allocate buffer for the JWT header and payload to be signed */
+	size_t msg_sz = strlen(hdr) + 1 + strlen(pay) + 1;
+	char *msg_out = calloc(msg_sz, 1);
+
+	/* Build the base64 URL JWT to sign:
+	 * <base64_header>.<base64_payload>
+	 */
+	if (msg_out) {
+		int ret = snprintk(msg_out, msg_sz, "%s.%s", hdr, pay);
+
+		if ((ret < 0) || (ret >= msg_sz)) {
+			LOG_ERR("Could not format JWT to be signed");
+			free(msg_out);
+			msg_out = NULL;
+		}
+	}
+
+	return msg_out;
+}
+
+//---------------------------------------------------------------------------//
+static char *jwt_payload_create(const char *const sub, const char *const aud, uint64_t exp)
+{
+	int err = 0;
+	char *pay_str = NULL;
+	cJSON *jwt_pay = cJSON_CreateObject();
+
+	// Issued At: format: time in seconds as integer
+	uint64_t issue_time = 0;
+
+	err = date_time_now(&issue_time);
```

Comment:
The date_time library updates the system clock. You could probably use `clock_gettime()` here instead.